### PR TITLE
fix(date): Validate based on minDate and maxDate

### PR DESCRIPTION
**Pull request by gabrielseco**
## 1. Description
Add validation for `minDate` and `maxDate` for inputType `date`

Issue tracker (Private): https://linear.app/remote/issue/RMT-542

Solves RMT-542

## 2. How to test this

I've created an internal [MR](https://gitlab.com/remote-com/employ-starbase/dragon/-/merge_requests/18246) where you can test that everything is working

### CODE REVIEW

**Comment by sandrina-p** [comment]
I'll review this by EOD.

File: src/tests/createHeadlessForm.test.js [hunk 1]
```diff
@@ -1035,6 +1035,62 @@ describe('createHeadlessForm', () => {
       );
     });
```
Comment (sandrina-p) [inline src/tests/createHeadlessForm.test.js:1036]: I'll review this today.

File: src/yupSchema.js [hunk 2]
```diff
@@ -73,13 +106,57 @@ const yupSchemas = {
...
+        .nullable()
+        .trim()
+        .matches(
+          /(?:\d){4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|1[0-9]|2[0-9]|3[0-1])/,
+          `Must be a valid date in ${DEFAULT_DATE_FORMAT.toLocaleLowerCase()} format. e.g. ${todayDateHint}`
+        )
+        .test('minDate', `The date must be ${minDate} or after`, (value) =>
+          validateMinDate(value, minDate)
+        )
+        .test('maxDate', `The date must be ${maxDate} or before`, (value) =>
```
Comment (sandrina-p) [inline src/yupSchema.js:121]: nit: Can you add a period (.) at the end of each sentence? I know we need to add it too in other error messages, but let's leave that for another PR.

File: src/tests/createHeadlessForm.test.js [hunk 3]
```diff
@@ -1035,6 +1035,62 @@ describe('createHeadlessForm', () => {
...
+            schema: expect.any(Object),
+            type: 'date',
+            maxLength: 10,
+            minDate: '1922-03-01',
+            maxDate: '2022-03-01',
+          },
+        ],
+      });
+
+      const fieldValidator = result.fields[0].schema;
```
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/tests/createHeadlessForm.test.js:1056]: Yes, you're right.

Give me some time and I'll ping you when I finish

File: src/yupSchema.js [hunk 4]
```diff
@@ -73,13 +106,57 @@ const yupSchemas = {
...
+          validateMaxDate(value, maxDate)
+        );
+    }
+
+    return string()
+      .nullable()
+      .trim()
+      .matches(
+        /(?:\d){4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|1[0-9]|2[0-9]|3[0-1])/,
```
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/yupSchema.js:155]: Thanks, I didn't figure out the other day how to compose things.

File: src/yupSchema.js [hunk 5]
```diff
@@ -73,13 +106,57 @@ const yupSchemas = {
...
-      `Must be a valid date in ${DEFAULT_DATE_FORMAT.toLocaleLowerCase()} format. e.g. ${todayDateHint}`
-    ),
+  date: ({ minDate, maxDate }) => {
+    if (minDate && maxDate) {
+      return string()
+        .nullable()
+        .trim()
+        .matches(
+          /(?:\d){4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|1[0-9]|2[0-9]|3[0-1])/,
+          `Must be a valid date in ${DEFAULT_DATE_FORMAT.toLocaleLowerCase()} format. e.g. ${todayDateHint}`
```
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/yupSchema.js]: Done!

File: src/yupSchema.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/yupSchema.js:106]: Done

File: src/tests/createHeadlessForm.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/tests/createHeadlessForm.test.js]: Done!

File: src/yupSchema.js (hunk 2 above)
Comment (gabrielseco) [reply src/yupSchema.js]: Done

File: src/yupSchema.js [hunk 6]
```diff
@@ -135,6 +212,10 @@ const getYupSchema = ({ inputType, ...field }) => {
     return yupSchemas.radioOrSelect(optionValues);
   }
 
+  if (inputType === 'date') {
```
[the remark this answers is not in the document]
Comment (gabrielseco) [reply src/yupSchema.js]: If we're going to support the format keyword can we leave the conditional with only format or I am missing something?

```
 if (field.format === 'date') {
}
```

[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/yupSchema.js]: Great point, yes, let's just use the `format` verification. Ideally, we should use native keywords for validation as much as possible!

File: src/tests/createHeadlessForm.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/createHeadlessForm.test.js]: extra: Could you refactor the test above too?

[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/createHeadlessForm.test.js]: nit: `.toEqual(undefined)` -> `.toBeUndefined()`

File: src/tests/createHeadlessForm.test.js [hunk 7]
```diff
@@ -1035,6 +1035,94 @@ describe('createHeadlessForm', () => {
...
+        birthdate: 'The date must be 1922-03-01 or after.',
+      });
+
+      expect(validateForm({ birthdate: '1922-03-01' })).toEqual(undefined);
+
+      expect(validateForm({ birthdate: '2021-03-01' })).toEqual(undefined);
+    });
+
+    it('support "date" field type with a maxDate', () => {
+      const { fields, handleValidation } = createHeadlessForm(schemaInputTypeDate);
```
Comment (sandrina-p) [inline src/tests/createHeadlessForm.test.js:1074]: nit: Could you refactor the `schemaInputTypeDate` to be a plain object without `JSONSchemaBuilder`?

Also, in the `mockDateInput`, the `maxLength` is not necessary as the format is enough to validate the value properly.

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@remoteoss/json-schema-form",
-  "version": "0.4.1-beta.0",
```
Comment (sandrina-p) [inline package.json:3]: Have you created a MR in our internal codebase to ensure the Storybook and E2E tests pass? We also need to update the docs to mention the `minDate` and `maxDate` custom keywords

Comment (gabrielseco) [reply package.json:3]: Yes, but I need to make sure I try the last changes of today

See [here](https://gitlab.com/remote-com/employ-starbase/dragon/-/merge_requests/18246)

Comment (sandrina-p) [reply package.json:3]: Yes, exactly

File: src/tests/createHeadlessForm.test.js (hunk 1 above)
Comment (gabrielseco) [reply src/tests/createHeadlessForm.test.js:1036]: Ready for a new review

File: src/yupSchema.js [hunk 9]
```diff
@@ -101,6 +101,13 @@ const yupSchemas = {
   date: ({ minDate, maxDate }) => {
     let dateString = string()
       .nullable()
+      .transform((value) => {
+        if (value === '') {
```
Comment (sandrina-p) [inline src/yupSchema.js:105]: Please leave a comment about why this is needed, similar to how we did in the radio. It can be as simple as

```js
// @BUG RMT-518 - Same reason to radioOrSelect above.
```

File: src/yupSchema.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/yupSchema.js:113]: Great!

File: src/tests/createHeadlessForm.test.js (hunk 1 above)
Comment (sandrina-p) [reply src/tests/createHeadlessForm.test.js:1036]: on it now 👀

Comment (sandrina-p) [reply src/tests/createHeadlessForm.test.js:1036]: Review done! Left a few comments here and in the internal docs, and this is ready to go!

File: src/tests/createHeadlessForm.test.js [hunk 10]
```diff
@@ -1009,30 +1009,112 @@ describe('createHeadlessForm', () => {
     });
 
     it('support "date" field type', () => {
-      const result = createHeadlessForm(schemaInputTypeDate);
+      const { fields, handleValidation } = createHeadlessForm(schemaInputTypeDate);
```
Comment (sandrina-p) [inline src/tests/createHeadlessForm.test.js:1012]: Approved, let's go! Thanks a lot Gabriel!

Just one last thing. The MR title can be improved to follow the conventional commits:

- use `date` not `dates` (date is the inputType name)
- the convention is `type(scope): body` not `type(scope) - body`
- The body doesn't need to be redundant.

```diff
- fix(dates) - fix date validation minDate and maxDate
+ fix(date): Validate based on minDate and maxDate
```

**Review by sandrina-p — APPROVED** [review]

File: src/tests/createHeadlessForm.test.js (hunk 10 above)
Comment (sandrina-p) [reply src/tests/createHeadlessForm.test.js:1012]: Also, as this is a fix, it's a patch version, not a minor. So it will be 0.4.2, not 0.5.0.

Comment (gabrielseco) [reply src/tests/createHeadlessForm.test.js:1012]: I'll release the patch after I merge this.

Thanks for the review and your patience.
